File: v201/src/shippeo_pod_upload_doc.js
const {
	SHIPMENT_HEADER_TABLE,
	SHIPMENT_HEADER_TABLE_STREAM_QLQ,
	SNS_TOPIC_ARN,
	SHIPPEO_USERNAME,
	SHIPPEO_PASSWORD,
	SHIPPEO_GET_DOC_URL,
	SHIPPEO_UPLOAD_DOC_URL,
	LOG_TABLE,
	SHIPPEO_GET_TOKEN_URL,
	SHIPPEO_GET_DOC_API_KEY,
	WT_WEBSLI_API_URL,
} = process.env;
const { get } = require("lodash");
const axios = require("axios");
const FormData = require("form-data");
const AWS = require("aws-sdk");
const dynamoDB = new AWS.DynamoDB.DocumentClient();
const sns = new AWS.SNS();
const sqs = new AWS.SQS();
const fs = require("fs");
const path = require("path");
const { exec } = require("child_process");
const request = require("request");

module.exports.handler = async (event, context, callback) => {
	let body;
	try {
		console.info("Event: \n", JSON.stringify(event));
		const records = get(event, "Records", []);
		console.log(
			"🚀 ~ file: shippeo_pod_upload_doc.js:27 ~ module.exports.handler= ~ records:",
			records
		);

		// Get Token
		const username = SHIPPEO_USERNAME;
		const password = SHIPPEO_PASSWORD;
		const basicAuth = getBasicAuth({ username, password });
		console.log(basicAuth);
		const url = SHIPPEO_GET_TOKEN_URL;
		const getTokenRes = await getToken({ url, token: basicAuth });
		const token = get(getTokenRes, "data.token");
		await Promise.all(
			records.map(async (record) => {
				console.log(SHIPMENT_HEADER_TABLE);
				body = JSON.parse(get(record, "body", ""));
				const FK_OrderNo = get(body, "Item.PK_OrderNo", "");
				const houseBillNo = get(body, "Item.Housebill", "");
				// const validHouseBillNo = await getIfHouseBillNumberValid({
				// 	FK_OrderNo,
				// });
				// console.log(
				// 	"🚀 ~ file: shippeo_pod_upload_doc.js:30 ~ records.map ~ validHouseBillNop:",
				// 	validHouseBillNo
				// );

				// if (!validHouseBillNo) {
				// 	console.log(`${houseBillNo} is not valid.`);
				// 	return;
				// }

				const getDocUrl = `${SHIPPEO_GET_DOC_URL}?docType=HOUSEBILL&housebill=${houseBillNo}`;
				const xApiKey = SHIPPEO_GET_DOC_API_KEY;
				console.log(
					"🚀 ~ file: shippeo_pod_upload_doc.js:57 ~ records.map ~ getDocUrl:",
					getDocUrl
				);
				console.log(
					"🚀 ~ file: shippeo_pod_upload_doc.js:57 ~ records.map ~ xApiKey:",
					xApiKey
				);
				// const xApiKey = "fIZpXhfGKQ42h6zIs7EUetiJd9yiAui7LlZxbkFh";
				// const getDocumentRes = await getDocument({ url: getDocUrl, xApiKey });
				// const docs = get(getDocumentRes, "getDocumentResponse.documents", []);

				const uploadToUrl = `${SHIPPEO_UPLOAD_DOC_URL}/${6986204}/files`;
				await uploadDocs({ uploadToUrl, token, houseBillNo });
			})
		);
	} catch (error) {
		console.error("Error : \n", error);
		return;
		try {
			const params = {
				Subject: "Error on shippeo-pod-upload-doc lambda",
				Message: `EDI alert reports Error in omni,\n lambda :shippeo-pod-upload-doc \n ERROR: ${error}`, // The message you want to send
				TopicArn: SNS_TOPIC_ARN, // The ARN (Amazon Resource Name) of your SNS topic
			};
			const data = await sns.publish(params).promise();
			console.log("Error notification Message sent:", data.MessageId);

			const queueUrl = SHIPMENT_HEADER_TABLE_STREAM_QLQ;
			const queueMessage = {
				QueueUrl: queueUrl,
				MessageBody: JSON.stringify(body),
			};
			console.log(
				"🚀 ~ file: shipment_header_table_stream_processor.js:31 ~ event.Records.forEach ~ queueMessage:",
				queueMessage
			);
			await sqs.sendMessage(queueMessage).promise();
		} catch (err) {
			console.log("Error while sending error notification message:", err);
		}
		return callback(null, { statusCode: 500 });
		// return callback(null, { statusCode: 500, body: JSON.stringify(error) });
	}
};

const getBasicAuth = ({ username, password }) => {
	const credentials = `${username}:${password}`;
	// Encode the credentials using base64
	const base64Credentials = Buffer.from(credentials).toString("base64");
	return `Basic ${base64Credentials}`;
};

const getToken = async ({ url, token }) => {
	const { data } = await axios.post(
		url,
		{},
		{
			headers: {
				"Content-Type": "application/json",
				Authorization: token,
			},
		}
	);
	return data;
};

const getDocument = async ({ url, xApiKey }) => {
	const { data } = await axios.get(url, {
		headers: {
			"Content-Type": "application/json",
			"x-api-key": xApiKey,
		},
	});
	return data;
};

const uploadDocs = async ({ uploadToUrl, token, houseBillNo }) => {
	const formData = new FormData();
	// docs.forEach(async (doc) => {
	// 	const { data: fileContent } = await axios.get(get(doc, "url"), {
	// 		responseType: "arraybuffer",
	// 	});
	// 	console.log(
	// 		"🚀 ~ file: shippeo_pod_upload_doc.js:146 ~ docs.forEach ~ fileContent:",
	// 		typeof fileContent
	// 	);
	// 	const fileName = get(doc, "filename");
	// 	console.log(
	// 		"🚀 ~ file: shippeo_pod_upload_doc.js:149 ~ docs.forEach ~ fileName:",
	// 		fileName
	// 	);
	// 	const bufferData = Buffer.from(fileContent);
	// 	const base64String = bufferData.toString("base64");
	// 	fs.writeFileSync(fileName, base64String);
	// 	formData.append("attachments[]", fileContent);
	// 	// formData.append("attachments[]", fs.createReadStream(fileName));
	// });

	const { b64str, api_status_code, Res } = await callWtRestApi(houseBillNo);
	console.log("🚀 ~ file: shippeo_pod_upload_doc.js:165 ~ uploadDocs ~ b64str:", b64str)
	const filePath = `/tmp/${houseBillNo}.pdf`;
	const fileData = Buffer.from(b64str, "base64");
	console.log(
		"🚀 ~ file: shippeo_pod_upload_doc.js:169 ~ uploadDocs ~ filePath:",
		filePath
	);

	fs.writeFileSync(filePath, fileData);

	const fileStat = fs.statSync(filePath);
	const contentLength = String(get(fileStat, "size", 0));

	const resp = await callExternalApiToUploadDoc({
		filePath,
		fileSize: contentLength,
		token,
		url: uploadToUrl,
	});

	// const file = fs.createReadStream(filePath, {
	// 	encoding: "base64",
	// });
	// console.log(
	// 	"🚀 ~ file: shippeo_pod_upload_doc.js:169 ~ uploadDocs ~ file:",
	// 	file
	// );

	// formData.append("attachments[]", file);
	// const fileStat = fs.statSync(filePath);
	// const contentLength = String(get(fileStat, "size", 0));
	// console.log(
	// 	"🚀 ~ file: shippeo_pod_upload_doc.js:173 ~ uploadDocs ~ contentLength:",
	// 	contentLength
	// );

	if (resp.status === 200) {
		await insertLog({housebillNumber: houseBillNo, data: resp.data});
	}
};

const insertLog = async ({ housebillNumber, data }) => {
	const params = {
		TableName: LOG_TABLE,
		Item: {
			pKey: housebillNumber,
			data,
		},
	};

	try {
		await dynamoDB.put(params).promise();
		console.log("Item inserted successfully");
	} catch (err) {
		console.error("Unable to insert item. Error:", err);
	}
};

function createDirectoryIfNotExists(directoryPath) {
	if (!fs.existsSync(directoryPath)) {
		fs.mkdirSync(directoryPath, { recursive: true });
		console.log(`Directory '${directoryPath}' created.`);
	} else {
		console.log(`Directory '${directoryPath}' already exists.`);
	}
}

function convertBase64ToPdf(base64String, filePath) {
	return new Promise((resolve, reject) => {
		const fileData = Buffer.from(base64String, "base64");
		fs.writeFileSync(filePath, fileData, function (err) {
			if (err) {
				console.error("Error converting Base64 to PDF:", err);
				reject(err);
			}
			resolve();
		});
	});
}

async function callWtRestApi(housebill) {
	try {
		const url = `${process.env.WT_WEBSLI_API_URL}/9980f7b9eaffb71ce2f86734dae062/housebill=${housebill}/doctype=HOUSEBILL`;

				const response = await axios.get(url);
		return {
			b64str: response.data.wtDocs.wtDoc[0].b64str,
			api_status_code: response.status,
			Res: response,
		};
	} catch (error) {
		console.error(
			`Error calling WT REST API for housebill ${housebill}:`,
			error
		);
		return "error";
	}
}

const callExternalApiToUploadDoc = async ({
	filePath,
	fileSize,
	token,
	url,
}) => {
	const formData = new FormData();
	formData.append(
		"attachments[]",
		fs.createReadStream(filePath, {
			encoding: "base64",
			autoClose: true,
			start: 0,
			end: Number(fileSize),
		})
	);
	try {
		const config = {
			method: "post",
			maxBodyLength: Infinity,
			url: url,
			headers: {
				...formData.getHeaders(),
				Authorization: `Bearer ${token}`,
				"Content-Type": "multipart/form-data",
				"Content-Length": fileSize,
			},
			data: formData,
		};
		console.log(
			"🚀 ~ file: shippeo_pod_upload_doc.js:188 ~ uploadDocs ~ config:",
			JSON.stringify(config)
		);
		const response = await axios.request(config);
		console.log("Response Data:", response.data);
		console.log("Status Code:", response.status);
		return { data: response.data, status: response.status };
	} catch (error) {
		if (error.response) {
			// The request was made and the server responded with a non-2xx status code
			console.log("Status Code:", error.response.status);
			console.log("Response Data:", JSON.stringify(error.response.data)); // This contains the error response data
			console.log("Response Headers:", error.response.headers);
		} else if (error.request) {
			// The request was made but no response was received
			console.error("No response received:", error.request);
		} else {
			// An error occurred while setting up the request
			console.error("Error:", error.message);
		}
		return false;
	}
};

// const

const getIfHouseBillNumberValid = async ({ FK_OrderNo }) => {
	const getFileDataParams = {
		TableName: process.env.SHIPMENT_FILE_TABLE,
		KeyConditionExpression: "FK_OrderNo  = :FK_OrderNo ",
		FilterExpression:
			"(FK_DocType = :FK_DocType1 OR FK_DocType = :FK_DocType2) AND CustomerAccess = :CustomerAccess",
		ExpressionAttributeValues: {
			":FK_OrderNo ": FK_OrderNo,
			":FK_DocType1": "HCPOD",
			":FK_DocType2": "POD",
			":CustomerAccess": "Y",
		},
	};
	console.log(
		"🚀 ~ file: shippeo_pod_upload_doc.js:109 ~ getHouseBillNumber ~ getFileDataParams:",
		getFileDataParams
	);
	const result = await dynamoDB.query(getFileDataParams).promise();
	console.log(
		"🚀 ~ file: shippeo_pod_upload_doc.js:125 ~ getIfHouseBillNumberValid ~ result:",
		result
	);
	return get(result, "Items", []).length > 0;
};

File: v201/src/shippeo_pod_upload_doc.py
import os
import base64
import logging
import boto3
import json
import requests
import io
from datetime import datetime, timedelta

sns = boto3.client('sns')
sqs = boto3.client('sqs')
dynamodb = boto3.client('dynamodb')

LOGGER = logging.getLogger()
LOGGER.setLevel(logging.INFO)

SHIPMENT_HEADER_TABLE = os.environ["SHIPMENT_HEADER_TABLE"]
SHIPMENT_HEADER_TABLE_STREAM_QLQ = os.environ["SHIPMENT_HEADER_TABLE_STREAM_QLQ"]
SNS_TOPIC_ARN = os.environ["SNS_TOPIC_ARN"]
SHIPPEO_USERNAME = os.environ["SHIPPEO_USERNAME"]
SHIPPEO_PASSWORD = os.environ["SHIPPEO_PASSWORD"]
SHIPPEO_GET_DOC_URL = os.environ["SHIPPEO_GET_DOC_URL"]
SHIPPEO_UPLOAD_DOC_URL = os.environ["SHIPPEO_UPLOAD_DOC_URL"]
LOG_TABLE = os.environ["LOG_TABLE"]
SHIPPEO_GET_TOKEN_URL = os.environ["SHIPPEO_GET_TOKEN_URL"]
SHIPPEO_GET_DOC_API_KEY = os.environ["SHIPPEO_GET_DOC_API_KEY"]
WT_WEBSLI_API_URL = os.environ["WT_WEBSLI_API_URL"]
SHIPMENT_FILE_TABLE = os.environ["SHIPMENT_FILE_TABLE"]
TOKEN_EXPIRATION_DAYS = os.environ["TOKEN_EXPIRATION_DAYS"]
TOKEN_VALIDATOR = os.environ["TOKEN_VALIDATOR"]
TOKEN_VALIDATION_TABLE_INDEX = os.environ["TOKEN_VALIDATION_TABLE_INDEX"]


def handler(event, context):
    LOGGER.info("Event: %s", event)
    records = event['Records']
    body = ""
    try:
        existing_token = get_existing_token()
        if existing_token == None:
            # Get token
            basic_auth = get_basic_auth(SHIPPEO_USERNAME, SHIPPEO_PASSWORD)
            LOGGER.info("basic_auth: %s", basic_auth)
            token_response = requests.post(
                SHIPPEO_GET_TOKEN_URL,
                headers={'Authorization': f'{basic_auth}'}
            )
            token = token_response.json()['data']['token']
            existing_token = token
            insert_token(token)

        websli_token = get_websli_token(SHIPPEO_GET_DOC_API_KEY)
        if websli_token == None:
            raise WebsliTokenNotFoundError("Websli token not found.")
        final_token = existing_token
        LOGGER.info("websli_token: %s", websli_token)
        for record in records:
            body = json.loads(record['body'])
            order_no = body['Item']['PK_OrderNo']
            housebill_no = body['Item']['Housebill']

            doc_type = get_if_house_bill_number_valid(order_no)
            LOGGER.info("doc_type: %s", doc_type)

            if doc_type == None:
                LOGGER.info(
                    f"House bill : {housebill_no} with order no: {order_no} is not valid")
                return {
                    'statusCode': 200
                }

            upload_to_url = f"{SHIPPEO_UPLOAD_DOC_URL}/{housebill_no}/files"
            # Upload docs
            result = upload_docs(upload_to_url, final_token,
                                 housebill_no, websli_token, doc_type)
            if result['status'] == 200:
                insert_log(housebill_no, result['data'])
            return {
                'statusCode': 200
            }

    except Exception as e:
        logging.error(f"Error: {e}")
        try:
            sns.publish(
                TopicArn=SNS_TOPIC_ARN,
                Subject="Error on shippeo-pod-upload-doc lambda",
                Message=f"Error in shippeo-pod-upload-doc lambda: {e}"
            )

            sqs.send_message(
                QueueUrl=SHIPMENT_HEADER_TABLE_STREAM_QLQ,
                MessageBody=json.dumps(body)
            )
        except Exception as e:
            logging.error(f"Error sending error notifications: {e}")
        return {
            'statusCode': 500
        }


def get_existing_token():
    try:
        # Define the query parameters
        params = {
            'TableName': LOG_TABLE,
            'KeyConditionExpression': "pKey = :pKey",
            'ExpressionAttributeValues': {
                ":pKey": {'S': 'token'}
            },
        }
        # Query the DynamoDB table
        response = dynamodb.query(**params)
        # Check if there are items in the result
        if len(response['Items']) > 0:
            return response['Items'][0]['data']['S']
        else:
            return None

    except Exception as e:
        LOGGER.info(f"Unable to insert item. Error: {e}")


def insert_token(data):
    two_days_from_now = datetime.utcnow() + timedelta(days=float(TOKEN_EXPIRATION_DAYS))
    two_days_from_now_iso = two_days_from_now.isoformat()
    params = {
        'TableName': LOG_TABLE,
        'Item': {
            'pKey': {'S': 'token'},
            'data': {'S': data},
            'expiration': {'S': two_days_from_now_iso}
        }
    }

    try:
        dynamodb.put_item(**params)
        LOGGER.info("Item inserted successfully")
    except Exception as e:
        LOGGER.info(f"Unable to insert item. Error: {e}")


def get_websli_token(apiKey):
    try:
        # Define the query parameters
        params = {
            'TableName': TOKEN_VALIDATOR,
            'IndexName': TOKEN_VALIDATION_TABLE_INDEX,
            'KeyConditionExpression': "ApiKey = :ApiKey",
            'ExpressionAttributeValues': {
                ":ApiKey": {'S': apiKey}
            },
        }
        # Query the DynamoDB table
        response = dynamodb.query(**params)
        # Check if there are items in the result
        if len(response['Items']) > 0:
            return response['Items'][0]['websli_key']['S']
        else:
            return None

    except Exception as e:
        LOGGER.info(f"Unable to insert item. Error: {e}")


def get_basic_auth(username, password):
    credentials = f"{username}:{password}"
    base64_credentials = base64.b64encode(
        credentials.encode('utf-8')).decode('utf-8')
    return f"Basic {base64_credentials}"


def upload_docs(upload_to_url, token, house_bill_no, websli_token, doc_type):
    try:
        b64_data = ""
        file_name = ""
        b64str_response = call_wt_rest_api(
            house_bill_no, websli_token, doc_type)
        if b64str_response == 'error':
            LOGGER.info('Error calling WT REST API')
            # handle error case
        else:
            b64_data = b64str_response['b64str']
            file_name = b64str_response['file_name']

        file_data = io.BytesIO(base64.b64decode(b64_data))

        # Create a dictionary for the form data
        files = {'attachments[]': (
            file_name, file_data, 'application/octet-stream')}

        headers = {
            'Authorization': f'Bearer {token}',
            'Accept': '*/*',
            'Accept-Encoding': 'gzip, deflate, br'
        }

        response = requests.post(upload_to_url, headers=headers, files=files)

        # Print response data and status code
        LOGGER.info(f"Response Data: {response.text}")
        LOGGER.info(f"Status Code: {response.status_code}")

        return {'data': response.text, 'status': response.status_code}
    except Exception as e:
        logging.error(f"Upload document error: {e}")
        return {'status': 400}


def call_wt_rest_api(housebill, websli_token, doc_type):
    try:
        url = f"{WT_WEBSLI_API_URL}/{websli_token}/housebill={housebill}/doctype={doc_type}"
        LOGGER.info(f"url: {url}")
        response = requests.get(url)

        if response.status_code == 200:
            data = response.json()
            b64str = data['wtDocs']['wtDoc'][0]['b64str']
            file_name = data['wtDocs']['wtDoc'][0]['filename']
            return {
                'b64str': b64str,
                'file_name': file_name
            }
        else:
            LOGGER.info(
                f"Error calling WT REST API for housebill {housebill}. Status Code: {response.status_code}")
            raise GetDocumentError(
                f"Error calling WT REST API for housebill {housebill}. Status Code: {response.status_code}")

    except Exception as error:
        LOGGER.info(
            f"Error calling WT REST API for housebill {housebill}:{error}")
        return 'error'


def insert_log(housebill_number, data):
    now = datetime.utcnow().isoformat()
    params = {
        'TableName': LOG_TABLE,
        'Item': {
            'pKey': {'S': housebill_number},
            'data': {'S': json.dumps(data)},
            'lastUpdateTime': {'S': now},
        }
    }

    try:
        dynamodb.put_item(**params)
        LOGGER.info("Item inserted successfully")
    except Exception as e:
        LOGGER.info(f"Unable to insert item. Error: {e}")


def get_if_house_bill_number_valid(order_no):
    try:
        # Define the query parameters
        params = {
            'TableName': SHIPMENT_FILE_TABLE,
            'KeyConditionExpression': "FK_OrderNo = :FK_OrderNo",
            'FilterExpression': "(FK_DocType = :FK_DocType1 OR FK_DocType = :FK_DocType2) AND CustomerAccess = :CustomerAccess",
            'ExpressionAttributeValues': {
                ":FK_OrderNo": {'S': order_no},
                ":FK_DocType1": {'S': "HCPOD"},
                ":FK_DocType2": {'S': "POD"},
                ":CustomerAccess": {'S': "Y"},
            },
        }

        # Query the DynamoDB table
        response = dynamodb.query(**params)
        LOGGER.info(f"response: {response}")
        # Check if there are items in the result
        items = response['Items']
        if len(items) > 0:
            return items[0]['FK_DocType']['S']
        else:
            return None
    except Exception as e:
        LOGGER.info(f"Unable to insert item. Error: {e}")


class GetDocumentError(Exception):
    pass


class WebsliTokenNotFoundError(Exception):
    pass


class HouseBillNotValidError(Exception):
    pass
